test/FastTests/Client/WhatChanged.cs

```diff
@@ -317,6 +319,126 @@ public void WhatChanged_Delete_After_Change_Value()
+        {
+            using (var store = GetDocumentStore())
+            {
+                const string docID = "d/1";
+                using (var session = store.OpenSession())
+                {
+                    var d = new Doc();
+                    session.Store(d, docID);
+                    var meta = session.Advanced.GetMetadataFor(d);
+
+                    meta["Test-A"] = new[] { "a", "a", "a" };
+                    meta["Test-C"] = new[] { "c", "c", "c" };
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var d = session.Load<Doc>(docID);
+                    var meta = session.Advanced.GetMetadataFor(d);
+
+                    meta["Test-A"] = new[] { "b", "a", "c" };
+
+                    var changes = session.Advanced.WhatChanged();
+                    Assert.True(changes.Count == 1);
+                    Assert.True(changes[docID].Length == 5);
+                    Assert.True(changes[docID][3].Change == DocumentsChanges.ChangeType.ArrayValueChanged);
+                    Assert.True(changes[docID][3].FieldName == "Test-A");
+                    Assert.True(changes[docID][4].Change == DocumentsChanges.ChangeType.ArrayValueChanged);
+                    Assert.True(changes[docID][4].FieldName == "Test-A");
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var d = session.Load<Doc>(docID);
+                    var meta = session.Advanced.GetMetadataFor(d);
+
+                    meta.Remove("Test-A");
+
+                    var changes = session.Advanced.WhatChanged();
+                    Assert.True(changes.Count == 1 && changes.Values.First()[0].Change == DocumentsChanges.ChangeType.RemovedField);
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    var d = session.Load<Doc>(docID);
+                    var meta = session.Advanced.GetMetadataFor(d);
+
+                    meta.Remove("Test-A");
+                    meta.Remove("Test-C");
+                    meta["Test-B"] = new[] { "b", "b", "b" };
+                    meta["Test-D"] = new[] { "d", "d", "d" };
+
+                    var changes = session.Advanced.WhatChanged();
+                    Assert.True(changes.Count == 1);
+                    Assert.True(changes[docID].Length == 7);
+                    Assert.True(changes[docID][0].Change == DocumentsChanges.ChangeType.RemovedField);
+                    Assert.True(changes[docID][0].FieldName == "Test-A");
+                    Assert.True(changes[docID][1].Change == DocumentsChanges.ChangeType.RemovedField);
+                    Assert.True(changes[docID][1].FieldName == "Test-C");
+                    Assert.True(changes[docID][5].Change == DocumentsChanges.ChangeType.NewField);
```

Comment:
I think that for completeness it's better to add asserts for the rest of the array indexes (2-4) for this case and all others in this PR

Reply:
array indexes (2-4) return wrong information:
https://issues.hibernatingrhinos.com/issue/RavenDB-13838

Reply:
@maximburyak Nice!